lib/elixir/lib/calendar.ex

```diff
@@ -265,6 +265,43 @@ defmodule Calendar do
   """
   @callback valid_time?(hour, minute, second, microsecond) :: boolean
 
+  @doc """
+  Parses the string representation for a time returned by `c:time_to_string/4`
+  into a time-tuple.
+  """
+  @callback parse_time(String.t()) ::
+              {:ok, {hour, minute, second, microsecond}}
+              | {:error, atom}
+
+  @doc """
+  Parses the string representation for a date returned by `c:date_to_string/3`
+  into a date-tuple.
+  """
+  @callback parse_date(String.t()) ::
+              {:ok, {year, month, day}}
+              | {:error, atom}
+
+  @doc """
+  Parses the string representation for a naive datetime returned by
+  `c:naive_datetime_to_string/7` into a naive-datetime-tuple.
+
+  The given string may contain a timezone offset but it is ignored.
+  """
+  @callback parse_naive_datetime(String.t()) ::
+              {:ok, {year, month, day, hour, minute, second, microsecond}}
+              | {:error, atom}
+
+  @doc """
+  Parses the string representation for a datetime returned by
+  `c:datetime_to_string/11` into a datetime-tuple.
+
+  The returned datetime must be in UTC. The original `utc_offset`
+  it was written in must be returned in the result.
+  """
+  @callback parse_utc_datetime(String.t()) ::
```

Comment:
this name makes me think that we can only "parse UTC date times" as in succeed exclusively for UTC timezone, and return error otherwise, e.g. `:non_utc_timezone` reason, same that `~U` returns. We of course handle non-UTC datetime strings here and we need them for `DateTime.from_iso8601`. Maybe the callback should be named just `parse_datetime` then? I think the current name is fine as is, especially because we do return the utc offset, but I thought I'd mention this anyway.

(Of course for all the other callbacks we have this nice property that, say, for all valid dates we basically have : `date == parse_date(date_to_string(date))` (I'm simplifying the actual interface here). This is true for all types other than non-UTC date times. Not that we can do anything about it, but perhaps worth repeating this point.)

Reply:
I like the symmetry of `parse_date`, `parse_time` and `parse_datetime`. And since an offset can be given this would still be a reasonable choice?

Reply:
So the `utc` is not about the input, but the fact the output must be a datetime in UTC.